clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3490,6 +3490,138 @@ void handleFailure(Throwable throwable) {
         return new DescribeDelegationTokenResult(tokensFuture);
     }
 
+    private static final class ListGroupsResults {
+        private final List<Throwable> errors;
+        private final HashMap<String, GroupListing> listings;
+        private final HashSet<Node> remaining;
+        private final KafkaFutureImpl<Collection<Object>> future;
+
+        ListGroupsResults(Collection<Node> leaders,
+                          KafkaFutureImpl<Collection<Object>> future) {
+            this.errors = new ArrayList<>();
+            this.listings = new HashMap<>();
+            this.remaining = new HashSet<>(leaders);
+            this.future = future;
+            tryComplete();
+        }
+
+        synchronized void addError(Throwable throwable, Node node) {
+            ApiError error = ApiError.fromThrowable(throwable);
+            if (error.message() == null || error.message().isEmpty()) {
+                errors.add(error.error().exception("Error listing groups on " + node));
+            } else {
+                errors.add(error.error().exception("Error listing groups on " + node + ": " + error.message()));
+            }
+        }
+
+        synchronized void addListing(GroupListing listing) {
+            listings.put(listing.groupId(), listing);
```

Comment:
Why do we need to use a map? All types of groups should share the same namespace, so we could simplify the data structure to a single `List<Object>`. I mean, errors and listings could be merged.

Reply:
The requests are sent to all nodes. I suppose there is a small possibility that more than one node responds for a particular group due to timing effects, and the map would allow just one entry per group ID to end up in the results. A similar pattern is used for the other uses of the ListGroups API.

Reply:
>  I suppose there is a small possibility that more than one node responds for a particular group due to timing effects,

That makes sense. for another, why does `ListGroupsResult` not use `map` structure?

Reply:
Because it matches the pattern established by `ListConsumerGroupsResult`. I went for consistency with the existing classes and methods.